src/lib/pubkey/hss_lms/hss.cpp

```diff
@@ -0,0 +1,418 @@
+   // The concatenated algorithm types for all layers
+   sk_size += hss_params().L().get() * (sizeof(LMS_Algorithm_Type) + sizeof(LMOTS_Algorithm_Type));
+   sk_size += m_hss_seed.size() + m_identifier.size();
+   return sk_size;
+}
+
+HSS_LMS_PrivateKeyInternal::HSS_LMS_PrivateKeyInternal(HSS_LMS_Params hss_params,
+                                                       LMS_Seed hss_seed,
+                                                       LMS_Identifier identifier) :
+      m_hss_params(std::move(hss_params)),
+      m_hss_seed(std::move(hss_seed)),
+      m_identifier(std::move(identifier)),
+      m_current_idx(0),
+      m_sig_size(HSS_Signature::size(m_hss_params)) {
+   BOTAN_ARG_CHECK(m_hss_seed.size() == m_hss_params.params_at_level(HSS_Level(0)).lms_params().m(),
+                   "Invalid seed size");
+   BOTAN_ARG_CHECK(m_identifier.size() == LMS_IDENTIFIER_LEN, "Invalid identifier size");
+}
+
+secure_vector<uint8_t> HSS_LMS_PrivateKeyInternal::sign(std::span<const uint8_t> msg) {
+   secure_vector<uint8_t> sig(HSS_Signature::size(hss_params()));
+   BufferStuffer sig_stuffer(sig);
+   sig_stuffer.append(store_be(hss_params().L() - 1));
+
+   std::vector<LMS_Tree_Node_Idx> q = derive_lms_leaf_indices_from_hss_index(reserve_next_idx(), hss_params());
+
+   // Derive LMS private keys and compute buffers
+   std::vector<LMS_PrivateKey> lms_key_at_layer;
+   std::vector<StrongSpan<LMS_Signature_Bytes>> out_lms_sig_buffer_at_layer;
+   std::vector<std::span<uint8_t>> out_child_pk_buffer_at_layer;
+   for(HSS_Level layer(0); layer < hss_params().L(); ++layer) {
+      // Generate key for current layer
+      const HSS_LMS_Params::LMS_LMOTS_Params_Pair& layer_params = hss_params().params_at_level(layer);
+      if(layer == HSS_Level(0)) {
+         lms_key_at_layer.push_back(hss_derive_root_lms_private_key());
+      } else {
+         lms_key_at_layer.push_back(
+            hss_derive_child_lms_private_key(layer_params, lms_key_at_layer.back(), q.at(layer.get() - 1)));
+         out_child_pk_buffer_at_layer.push_back(sig_stuffer.next(LMS_PublicKey::size(layer_params.lms_params())));
+      }
+      out_lms_sig_buffer_at_layer.push_back(sig_stuffer.next<LMS_Signature_Bytes>(
+         LMS_Signature::size(layer_params.lms_params(), layer_params.lmots_params())));
+   }
+   BOTAN_ASSERT_NOMSG(sig_stuffer.full());
+
+   // Sign and write the signature from bottom layer to root layer
+   std::vector<uint8_t> current_pk;
+   for(int32_t layer_it = hss_params().L().get() - 1; layer_it >= 0; --layer_it) {
+      HSS_Level layer(layer_it);
+      if(layer == hss_params().L() - 1) {
+         current_pk =
+            lms_key_at_layer.at(layer.get())
+               .sign_and_get_pk(out_lms_sig_buffer_at_layer.at(layer.get()), q.at(layer.get()), LMS_Message(msg))
+               .to_bytes();
+      } else {
+         std::copy(current_pk.begin(), current_pk.end(), out_child_pk_buffer_at_layer.at(layer.get()).begin());
```

Comment:
```suggestion
         copy_mem(out_child_pk_buffer_at_layer.at(layer.get()), current_pk);
```

... didn't try this locally, but should work and runtime-check the range sizes.